### PR TITLE
IrGraphGenerator: export fuser IR to graphviz

### PR DESCRIPTION
**Pull request by tlemo**
Generate graphviz graphs representing fuser IRs. See IrGraphGenerator comments for basic usage and tips.

The generated graphs are saved as local files (the convention is to use .dot extension), which can be used with any graphviz tool:
- xdot 
- http://viz-js.com
- http://graphviz.it

For VS Code users, the tintinweb.graphviz-interactive-preview is highly recommended.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// Notes:
+//  - When called from the debugger, the detail_level must be
+//    explicitly passed in (most debuggers don't support default arguments)
+//
+//  - The output dot file path can't include shell specific notations,
+//    for example you can't use "~/temp/ir.dot" ("/home/user/temp/ir.dot"
+//    must be used instead)
+//
+class TORCH_CUDA_API IrGraphGenerator : private OptInConstDispatch {
+  friend void torch::jit::testGPU_IrGraphGenerator();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_graphviz.h]: I have a slight preference for dedicated testing interfaces, which would avoid this ugliness, but I'm aware that not everyone is comfortable with public testing interfaces.

Any thoughts? I'd be more than happy to make the switch if there are no objections.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+ public:
+  enum class DetailLevel {
+    ComputeOnly, // Only dataflow (compute) nodes
+    Basic, // Compute + schedule, with minimal details (default)
+    Explicit, // Additional details (ex. symbolic names for scalar constants)
+    Verbose, // Includes all values and dead definitions
+  };
+
+ public:
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_graphviz.h:45]: nitpick `public:` here again.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  void addArc(
+      const Statement* src,
+      const Statement* dst,
+      const std::string& style = "");
+
+  void printExpr(const Expr* expr, const std::string& label);
+  void printValue(const Val* val, const std::string& label);
+
+ private:
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_graphviz.h:101]: nitpitck: `private:` again.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.h:101]: NVM, didn't see the it's the variable section....

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h (hunk 2 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.h:45]: public is technically redundant here indeed, although I prefer to put member types in a separate section from member functions (and member data) sections.

is this frowned upon?

File: torch/csrc/jit/codegen/cuda/ir_graphviz.cpp [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+      const auto root_domain = tv->getRootDomain();
+      addArc(tv, root_domain, "[style=dashed, color=green, arrowhead=none]");
+    }
+  }
+
+  graph_def_ << "  }\n";
+}
+
+void IrGraphGenerator::handle(const Statement* s) {
+  OptInConstDispatch::handle(s);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:285]: If we are not really overriding the `handle(const Statement* s)` from inherited class, maybe we don't want to have a dummy implementation here.
You should be able to get it working with explicit `using OptInConstDispatch::handle` in the class declaration.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.cpp [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+};
+
+void IrGraphGenerator::handle(const Val* v) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:288]: nitpick.
I think if we use the Dispatch thing, you would want to go through `handle(Statement/Val/Expr)` via definition in inherited-scope and optin/out handles for the rest of the nodes.
But I see how it could save code in your implemented, so I'm totally fine with it if you decide to leave it as is.

However, if we are re-implementing all these stock functions from the inherited class, I don't see much value in inheritance here, especially that we now have branch to skip the call to inherited handle function.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.h]: IIUC, the only thing that we are checking in the tests are generated strings. If we slightly change the public `print` function to output string, this is not necessary any more, right?

On related topic, only being able to write output to a file seems quite awkward to me. It would be easier let user opt-in the output format. In that case we can opt for output as stringstream and test it that way. So we don't need to declare friend here.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.h:45]: If this is intentional, then totally ignore me.
I don't have any opinion on this and was pointing it out just in case it's an oversight.

**Review by csarofeen — CHANGES_REQUESTED** [review]
I don't want any full graph passes that aren't iterative. 
Everything else looks good or was mentioned by Jie.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.h (hunk 1 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.h]: The intention is to make the IR printing as convenient as possible for debugging - so one call either from the code or from the debugger.

The "return string" is exactly the testing interface I had in mind, so we're probably talking about the same thing (I tried to resist adding it unless there's a good justification for it)

I'll make the changes.

File: torch/csrc/jit/codegen/cuda/ir_graphviz.cpp (hunk 4 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:285]: You're right, using OptInConstDispatch::handle would do here, but I prefer the symmetry with the other overrides. 

Not only that, but "using" is a blunt tool: it brings the whole overload set into view, not just a specific one (you lose the benefit of explicit overrides and it can lead to unpredictable issues if the base class interface is changing)

File: torch/csrc/jit/codegen/cuda/ir_graphviz.cpp (hunk 5 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:288]: > via definition in inherited-scope and optin/out handles for the rest of the nodes",
I'm not sure I follow, can you please elaborate?

> especially that we now have branch to skip the call to inherited handle function.
Maybe I didn't parse this right, which "skip the call" are you talking about?

**Comment by tlemo** [comment]
> I don't want any full graph passes that aren't iterative.

This is a debug tool, so at most it would demonstrate the limitations of the recursive approach. 

Moreover, as a low level debugging tool there's some value in having separate iteration code path (as in providing a bit of cross checking and it also avoids adding extra requirements on a generic production traversal iterator)

File: torch/csrc/jit/codegen/cuda/ir_graphviz.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:288]: Sorry for poor English :bow: 
> via definition in inherited-scope and optin/out handles for the rest of the nodes

This is basically what I was suggesting in the other comment, take the big hammer of `using OptInConstDispatch::handle` and get rid of the few `handle(Statement/Val/Expr)`.

> skip the call to inherited handle function.

I'm talking about the if statement in the function. We only calls `OptInConstDispatch::handle` when current node has not been `visited`. This is different from how inherited class traverses nodes.
https://github.com/csarofeen/pytorch/pull/44/files/4d96c2c534a641ab161d3f3bfb958e0848853b7c#diff-83aa246d49f3b1bca5f60e92bcaf616bR291
Basically inherited class only serves the purpose of an abstract interface. I guess I still see value here, as they would raise error when new nodes are added to codebase but not to `IrGraphGenerator`.

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_graphviz.cpp:288]: As you noted, handing Val/Expr are not pass-through to the base implementation, so using would only save us one override: handle(Statement*)

> Basically inherited class only serves the purpose of an abstract interface. I guess I still see value here, as they would raise error when new nodes are added to codebase but not to IrGraphGenerator.

Sorry, I'm still not sure I understand: IrGraphGenerator implements the OptInConstDispatch for dispatching (which is a hidden part of its implementation, thus the private inheritance).
How would we do the dispatch w/o inheritance?

**Comment by csarofeen** [comment]
It looks almost exactly like: https://github.com/csarofeen/pytorch/blob/master/torch/csrc/jit/codegen/cuda/ir_iostream.h

I don't see how it would impose any new requirements. Traversal should be checked carefully, but not by trying to compare with a debug tool. I also think the usefulness of this will and can go farther than a debug utilities, primarily an automated reporting mechanism from users. I actually think consistency would get us further in this case. I assume we would want to know how exactly what we're iterating over, not another potentially different representation.

**Review by csarofeen — APPROVED** [review]
